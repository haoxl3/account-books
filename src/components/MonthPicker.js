import React from 'react'
import PropTypes from 'prop-types'
import {padLeft, range} from '../utility'

class MonthPicker extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            isOpen: false,
            selectedYear: this.props.year
        }
    }
    // 如果不想使用bind，可使用() => {}代替，默认在create-react-app中开启
    toggleDropdown = (event) => {
        event.preventDefault()
        this.setState({
            isOpen: !this.state.isOpen
        })
    }
    // 关闭下拉框，并回传选择的年月
    selectYear = (event, yearNumber) => {
        event.preventDefault()
        this.setState({
            selectedYear: yearNumber
        })
    }
    // 高亮当前点击的月份
    selectMonth = (event, monthNumber) => {
        event.preventDefault()
        this.setState({
            isOpen: false
        })
        this.props.onChange(this.state.selectedYear, monthNumber)
    }
    render() {
        const {year, month} = this.props
        const {isOpen} = this.state
        const {selectedYear} = this.state
        const monthRange = range(12, 1)
        const yearRange = range(9, -4).map(number => number + year)
        return (
            <div className="dropdown month-picker-component">
                <h4>选择月份</h4>
                <button 
                    className="btn btn-lg btn-secondary dropdown-toggle"
                    onClick={this.toggleDropdown}
                >
                    {`${year}年 ${padLeft(month)}月`}
                </button>
                { isOpen &&
                    <div className="dropdown-menu" style={{display:'block'}}>
                        <div className="row">
                            <div className="col border-right">
                                { yearRange.map((yearNumber, index) => 
                                    <a 
                                        key={index} 
                                        className={(yearNumber === selectedYear) ? 'dropdown-item active':'dropdown-item'}
                                        onClick={(event) => {this.selectYear(event, yearNumber)}}
                                    >
                                        {yearNumber}年
                                    </a>
                                )}
                            </div>
                            <div className="col">
                                { monthRange.map((monthNumber, index) => 
                                    <a 
                                        key={index} 
                                        className={(monthNumber === month) ? 'dropdown-item active':'dropdown-item'}
                                        onClick={(event) => {this.selectMonth(event, monthNumber)}}
                                    >
                                        {padLeft(monthNumber)}月
                                    </a>
                                )}
                            </div>
                        </div>
                    </div>
                }
            </div>
        )
    }
}
MonthPicker.propTypes = {
    year: PropTypes.number.isRequired,
    month: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired
}
export default MonthPicker