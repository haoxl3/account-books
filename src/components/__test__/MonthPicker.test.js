import React from 'react'
import ReactDOM from 'react-dom'
import {mount} from 'enzyme'
import MonthPicker from '../MonthPicker'

let props = {
    year: 2018,
    month: 8,
    onChange: jest.fn()
}
let wrapper

describe('test MOnthPicker component', () => {
    beforeEach(() => {
        wrapper = mount(<MonthPicker {...props}/>)
    })
    it('after the dropdown is shown, click the document should close the', () => {
        let eventMap = {}
        document.addEventListener = jest.fn((event, cb) => {
            eventMap[event] = cb
        })
        const wrapper = mount(<MonthPicker {...props}/>)
        wrapper.find('.dropdown-toggle').simulate('click')
        expect(wrapper.state('isOpen')).toEqual(true)
        expect(wrapper.find('.dropdown-menu').length).toEqual(1)
        // 找到dom节点
        eventMap.click({
            target: ReactDOM.findDOMNode(wrapper.instance())
        })
        expect(wrapper.state('isOpen')).toEqual(true)

        eventMap.click({
            target: document
        })
        expect(wrapper.state('isOpen')).toEqual(false)
    })
})