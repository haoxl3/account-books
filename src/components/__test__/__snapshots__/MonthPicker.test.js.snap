// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MOnthPicker component should render the component to match the snapshot 1`] = `undefined`;

exports[`test MonthPicker component should render the component to match the snapshot 1`] = `undefined`;
